src/engine/strat_engine/pool.rs

```diff
@@ -309,6 +309,12 @@ impl Pool for StratPool {
         paths: &[&Path],
         tier: BlockDevTier,
     ) -> StratisResult<SetCreateAction<DevUuid>> {
+        if paths.is_empty() {
```

Comment:
The rest of the semantic changes seem right, but this one seems like it's too broad. In the general case, I think the empty list should be viewed as a set that is already contained in the blockdevs already belonging to the tier, whether it is cache or data. The error case should be restricted to where this would kick off making a new cache, and that can now be checked w/ the has_cache() function on ```Pool```. So something like:

```
if paths.is_empty() {
  return if !self.has_cache() && tier == Cache {
    Err(...)  
  } else {
    Ok(Identity)
  }
}

Reply:
I see your point, and I am still concerned that this will be a very confusing thing for people to navigate. From a mathematical perspective, this makes sense, but from a user standpoint, I'm not sure this will be clear to those who are not thinking in terms of set theory. I don't think everyone using Stratis will have that background. If we do decide to go with this option, I think we need to be very clear in the error message about why it's failing because up until this point, we have not introduced any idempotent behavior that is accepted in one context where the action has already been done and fails when the action is requested and could be acted upon. This seems inconsistent with the line we've perhaps accidentally drawn and I'd like to be very clear with the user about why we're making this exception if you're set on this being the way forward.

Reply:
> I see your point, and I am still concerned that this will be a very confusing thing for people to navigate. From a mathematical perspective, this makes sense, but from a user standpoint, I'm not sure this will be clear to those who are not thinking in terms of set theory. I don't think everyone using Stratis will have that background. If we do decide to go with this option, I think we need to be very clear in the error message about why it's failing because up until this point, we have not introduced any idempotent behavior that is accepted in one context where the action has already been done and fails when the action is requested and could be acted upon. This seems inconsistent with the line we've perhaps accidentally drawn and I'd like to be very clear with the user about why we're making this exception if you're set on this being the way forward.

I'm not sure about the meaning one of " we have not introduced any dempotent behavior that is accepted in one context where the action has already been done and fails when the action is requested and could be acted upon". But remember, we're not asking users of Stratis to understand set theory, only people who write clients of the D-Bus interface. Defining users of Stratis as people who use the CLI, they'll never see this because the parser will reject commands that specify 0 blockdevs.